Cross-platform CI scripts

aminya opened the pull request:
### Description of the change
- This uses Powershell 7 as a cross-platform shell for CI scripts:
	- PowerShell 7 allows using the same syntax and code across all the operating systems.
	- This simplifies CI scripts. For example, previously CMD was used in Windows, which is an old shell with limited functionality.
	- The old linux_trusty container did not detect pwsh, so we use the Linux image provided by Microsoft, and we use Clang 9 that is provided in it.

- This refactors CI scripts into separate files (templates): 
	- It also allows using parallelization of the tests without copy-pasting the same code. For example: 
		- In macOS, the same templates are reused in the test phase. 
		- In Windows x86, the templates are used for parallelization of tests in [this PR](https://github.com/atom-ide-community/atom/pull/31)

	- This makes it easier to manage CI code. For example, we can quickly compare the publish step for all operating systems without going through big files and trying to find the relevant steps.

Previously, the CI scripts were written back in ~2017. At that time, Azure Templates had not been introduced yet, and because of that using PowerShell was not suitable yet. Now in 2020, we can benefit from these and improve the CI. This is also a requirement for the future Test Parallelization pull request.

### Verification
The CI passes. Tested in atom-ide-community

### Release Notes
N/A

~### Includes~
~This includes #21023.~
~This supersedes #21057.~

~### Thing you should do~
~Because this includes #21023, it requires you to set `SHOULD_SIGN=true` as an environment variable inside Azure.~

### Fork Pull request
https://github.com/atom-ide-community/atom/pull/46

aminya committed:
using templates in nightly-release.yml

aminya committed:
using templates in release-branch-build.yml

DeeDeeG committed:
CI: Delete outdated comment in platforms/linux.yml

We no-longer run the Linux build in an old Ubuntu 14.04 "Trusty Tahr"
Docker container. We run builds directly in a virtualized Ubuntu 16.04
"Xenial Xerus" image now.

DeeDeeG committed:
CI: Move GetReleaseVersion to its own template

DeeDeeG committed:
CI: Add NightlyFlag parameter in GetReleaseVersion

We can use this to run `get-release-version.js`
with the `--nightly` flag on the "Atom Nightly" pipeline.

sadick254 commented:
@aminya Great job 👍 

Since this is a huge refactor I would recommend splitting it into multiple PR's(1 for each template). That way it becomes easy to review and track the impact the changes have on our CI.

Am happy to assist with that if need be.

aminya commented:
Start by merging https://github.com/atom/atom/pull/21023. The commits are granular, so they can be grouped together and cherry-picked if needed.

sadick254 commented:
@aminya I see this now has conflicts that need to be resolved.

aminya commented:
> @aminya I see this now has conflicts that need to be resolved.

I rebased this! The next step is merging #21057

sadick254 commented:
@aminya https://github.com/atom/atom/pull/21057 has been merged. You might need to rebase again.

aminya commented:
@sadick254 Thanks! I cherry-picked the preparation template. The next step is merging https://github.com/atom/atom/pull/21081!

aminya commented:
@sadick254 Thanks! I cherry-picked the preparation template. The next step is merging https://github.com/atom/atom/pull/21221

aminya commented:
@sadick254 This is rebased again. The next one is https://github.com/atom/atom/pull/21226

aminya commented:
@sadick254 

This is rebased again. The next one is the build template. 
https://github.com/atom/atom/pull/21230

aminya commented:
@sadick254 

This is rebased again. The next one is the build template. 
https://github.com/atom/atom/pull/21230

aminya commented:
@sadick254 The next one is #21262

aminya committed:
rename "buildArch" to "BUILD_ARCH"

Co-authored-by: Amin Yahyaabadi <aminyahyaabadi74@gmail.com>

aminya commented:
@sadick254 This is rebased. I think this can go in one PR. It has the test-template + miscellaneous changes (in the last 4 commits).

Let me know if you want me to break this into two PR.

aminya committed:
CI: Don't upgrade npm twice on Windows

We already install the npm version we want in previous steps

https://github.com/atom-ide-community/atom/pull/140

Co-Authored-By: DeeDeeG <DeeDeeG@users.noreply.github.com>

sadick254 commented:
@aminya Breaking it to two PR's would be great.

sadick254 commented:
This needs another rebasing. @aminya

aminya committed:
CI: Drop old workaround, refactor script blocks

This old bug in the CI inexplicably went away at some point.

Co-Authored-By: DeeDeeG <DeeDeeG@users.noreply.github.com>

aminya commented:
@sadick254 Done! Now, this can be merged.

aminya commented on script/vsts/platforms/templates/get-release-version.yml:
```diff
@@ -0,0 +1,19 @@
+  steps:
+    - script: |
+        cd script/vsts
+        npm ci
```
I noticed that this is `npm ci` which reverts your change to npm install change. @sadick254.

DeeDeeG replied:
Very briefly: I'd advocate for using `npm ci`.

---

@sadick254 If I recall correctly, the `npm ci` --> `npm install` change was meant to fix intermittent Linux build failures.

It turned out `npm install` doesn't fix those Linux failures. Node 12.16.3 in CI seems to be doing okay for fixing the Linux build failures. (So there isn't a true purpose left as for why to prefer `npm install` over `npm ci`, as I understand it.)

In my personal view, it is slightly advantageous to use `npm ci` iwherever we can. Just a bit faster and cleaner than `npm install`. I admit, though, the speedup is only a few seconds most of the time.

DeeDeeG replied:
But don't worry about that if it slows down merging this. The stakes are so low as to not really matter. To be quite honest.

aminya committed:
use npm install in get-release-version

https://github.com/atom/atom/pull/21081/files#r487132495

aminya commented:
@sadick254 Done! Now, this can be merged.

sadick254 reviewed (approved):
This looks great. Thanks @aminya and @DeeDeeG for the great collaboration on this one. 👍
